text/0000-pmgr.md

```diff
@@ -0,0 +1,257 @@
+
+## Setup
+
+The kernel expects the init program to communicate to it a pair of object IDs that are each Queue
+objects---the *kernel request queue*, and the *pager request queue*. These objects allow for
+communication between the kernel and userspace, where the userspace portion is expected to be a
+binary program started by init that handles kernel paging requests.
+
+This coordination on which objects to use for queues is done via the kernel using the `sys_new_handle` system call. The
+init program calls this function twice to record first the kernel request queue, and second the pager request queue.
+Once these queues' IDs are both told to the kernel, the kernel starts its internal paging mechanisms (implementation details not
+outlined in this RFC).
+
+The kernel pager request queue's submission and completion types are:
+
+```{rust}
+enum KernelRequest {
+    ObjectInfoReq(ObjID),
+    PageDataReq(ObjID, [ObjectRange; A]),
+    DramRel(usize),
+    DramPages([PhysRange; B]),
+    Evict(ObjID, [ObjectRange; C]),
+    Sync(ObjID, [ObjectRange; C]),
+}
+
+enum KernelCompletion {
+    Success,
+    ObjectInfo(ObjectInfo),
+    PageInfo([PhysRange; D]),
+    DramPages([PhysRange; E]),
+    Err(PagerReqError),
+}
+```
+
+The pager's request queue has the following types:
+
+```
+enum PagerRequest {
+    PageInfo([(PhysRange, ObjectRange); F]),
+    ObjectInfo(ObjectInfo),
+    DramReq(usize),
+    DramPages([PhysRange; G]),
+    Evict(ObjID, [ObjectRange; H])
+}
+
+enum PagerCompletion {
+    Success,
+    DramPages([PhysRange, I]),
+    Err(KernelReqError),
+}
+```
+
+The pager can directly submit information about pages and objects to the kernel before the kernel
+even asks for it, allowing for prefetching (possibly in the background, or application-driven). The
+kernel can choose to ignore these prefetching requests.
+
+## Memory Management
+
+To allow for the pager to implement some storage mechanism that does DMA transfer, the kernel allows
+for DRAM-related requests over these queues. At any time the kernel can offer up unused pages of
+DRAM for the pager to gain ownership over, and the pager can also issue a request for additional
+pages. When requesting additional pages, the kernel is allowed to refuse if low on memory. The
+kernel can also issue a request to the pager to release DRAM pages. The pager is allowed to refuse.
+For this reason, the kernel should not pass ownership of all DRAM to the pager.
+
+### Access to Physical Memory
+
+Since the pager exists in userspace, it does not have access to the DRAM pages that are given to it
+by the kernel. While in most operation, it does not need to read DRAM pages (the pager can submit
+DMA commands without access to the memory), occasionally it might want to read page contents. To do
+this, the pager can simply invent an object ID and submit PageInfo requests to the kernel to
+temporarily map pages into accessable memory.
```

Comment:
[question] is this the only way to do this? It feels hacky (which might be fine), but isn't there also a (small) chance of ID collision, depending on how the pager comes up with the nonsense object ID it submits to the kernel?

Reply:
The pager can either use the low-collision-probability ID generation scheme we have already, or we could define object IDs with the upper 64-bits set to zero as "reserved" for things like this.